streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java

```diff
@@ -16,71 +16,186 @@
+        suppressDurationMillis = suppress.getTimeToWaitForMoreEvents().toMillis();
+        buffer = new InMemoryTimeOrderedKeyValueBuffer<>();
+        timeDefinition = suppress.getTimeDefinition();
+        bufferFullStrategy = suppress.getBufferConfig().bufferFullStrategy();
+        shouldSuppressTombstones = suppress.shouldSuppressTombstones();
     }
 
     @Override
     public void init(final ProcessorContext context) {
         internalProcessorContext = (InternalProcessorContext) context;
+        this.keySerde = keySerde == null ? castSerde(context.keySerde()) : keySerde;
+        this.valueSerde =
+            valueSerde == null ? new FullChangeSerde<>(castSerde(context.valueSerde())) : valueSerde;
     }
 
     @Override
     public void process(final K key, final Change<V> value) {
-        if (suppress.getTimeToWaitForMoreEvents() == Duration.ZERO && definedRecordTime(key) <= internalProcessorContext.streamTime()) {
+        if (suppressDurationMillis == 0 && definedRecordTime(key) <= internalProcessorContext.streamTime()) {
             if (shouldForward(value)) {
                 internalProcessorContext.forward(key, value);
             } // else skip
         } else {
-            throw new NotImplementedException();
+            buffer(key, value);
+            enforceConstraints(internalProcessorContext.streamTime());
         }
     }
 
-    private boolean shouldForward(final Change<V> value) {
-        return !(value.newValue == null && suppress.suppressTombstones());
+    @SuppressWarnings("unchecked")
+    private <T> Serde<T> castSerde(final Serde<?> untyped) {
+        return (Serde<T>) untyped;
+    }
+
+    private void buffer(final K key, final Change<V> value) {
+        final long time = definedRecordTime(key);
+        final ProcessorRecordContext recordContext = internalProcessorContext.recordContext();
+
+        buffer.put(time, key, new ContextualRecord<>(time, value, recordContext), computeRecordSize(key, value, recordContext));
     }
 
     private long definedRecordTime(final K key) {
-        return suppress.getTimeDefinition().time(internalProcessorContext, key);
+        return timeDefinition.time(internalProcessorContext, key);
     }
 
-    @Override
-    public void close() {
+    private void enforceConstraints(final long streamTime) {
+        final long expiryTime = streamTime - suppressDurationMillis;
+        if (overCapacity() || buffer.minTimestamp() <= expiryTime) {
+
+            try (final AutoCloseableIterator<KeyValue<TimeKey<K>, ContextualRecord<Change<V>>>> iterator = buffer.iterator()) {
+
+                // drainExpiredRecords may have called iterator.next,
+                // in which case, we should consume that value before calling next again
+                final KeyValue<TimeKey<K>, ContextualRecord<Change<V>>> maybeRemaining =
+                    drainExpiredRecords(expiryTime, iterator);
+
+                if (overCapacity()) {
```

Comment:
Can we add two cases to `KTableSuppressProcessorTest` to hit this branch? One for the `EMIT` case and another for the `SHUT_DOWN`